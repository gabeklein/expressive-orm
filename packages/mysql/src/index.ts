export {
  default as Connection,
  default as MySQLConnection
} from './Connection';

export * from "@expressive/sql";