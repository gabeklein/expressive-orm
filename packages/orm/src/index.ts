export { default, Type } from './Type';
export type { Compat, Insert, Values } from './Type';
export * from './fields';
export * from './where';
export { one } from './one';
export { get } from './get';