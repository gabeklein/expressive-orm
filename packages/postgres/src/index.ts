export * from "@expressive/sql";
export { PGLiteConnection } from "./PGLiteConnection";
export { PGConnection } from "./PGConnection";