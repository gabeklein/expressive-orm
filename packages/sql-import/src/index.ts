export { Scanner } from './Scanner';
export { Parser } from './generate/Parser';