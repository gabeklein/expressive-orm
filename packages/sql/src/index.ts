export { Bool } from './field/Bool';
export { Connection } from './connection/Connection';
export { default as Util } from './utility';
export { Enum } from './field/Enum';
export { Field } from './Field';
export { Join } from './field/Join';
export { Json } from './field/Json';
export { Many } from './field/Many';
export { Num } from './field/Num';
export { One } from './field/One';
export { Query } from './query/Query';
export { Ref } from './field/Ref';
export { Scanner } from './parse/Scanner';
export { Schema } from './connection/Schema';
export { Str } from './field/Str';
export { Table } from './field/Table';
export { Time } from './field/Time';
export { Type } from './Type';