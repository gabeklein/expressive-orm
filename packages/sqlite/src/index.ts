import SQLiteConnection from './Connection';

export * from "@expressive/orm";

export {
  SQLiteConnection,
  SQLiteConnection as default
}