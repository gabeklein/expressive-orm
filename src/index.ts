import Bool from './columns/Bool';
import Int from './columns/Int';
import Many from './columns/Many';
import One from './columns/One';
import VarChar from './columns/VarChar';
import Entity from './entity';

class Author extends Entity {
  name = VarChar();
  age = Int();
  nickname = VarChar({ nullable: true });
  active = Bool();
  books = Many(Book);
}

class Book extends Entity {
  title = VarChar();
  author = One(Author); 
}

Author.register();
Book.register();

export async function getBook(){
  return Book.getOne({
    where(){
      this.title.is("1984");
      this.author.age.isMore(50);
    },
    select(){
      return this.title;
    }
  })
}

export async function getAuthor(){
  return Author.getOne({
    where(){
      this.name.isNot("gabe");
      this.nickname.is("gabe");
      this.age.isMore(3);
      this.active.is(true);
    },
    select(){
      return {
        name: this.nickname,
        age: this.age,
        status: this.active ? "active" : "inactive"
      } as const
    }
  });
}

console.clear()
getBook();