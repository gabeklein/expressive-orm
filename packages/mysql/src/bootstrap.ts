import { Field, Type, Util } from '@expressive/sql';

export function bootstrap(entities: Iterable<Type.EntityType>){
  const commands = [] as string[];
  
  for(const entity of entities)
    entity.ready();

  for(const entity of entities)
    commands.push(create(entity));

  for(const entity of entities){
    const statement = constraints(entity);

    if(statement)
      commands.push(statement);
  }
  
  return commands.join(";");
}

export function drop(table: Type.EntityType){
  return `DROP TABLE IF EXISTS ${table.name}`
}

export function create(table: Type.EntityType){
  const { table: name } = table;
  const statements = [] as string[];

  table.fields.forEach(field => {
    const sql = column(field);

    if(sql)
      statements.push(sql);
  });

  return `CREATE TABLE IF NOT EXISTS ${name} (${statements.join(",")})`
}

export function constraints(table: Type.EntityType){
  const statement = [] as string[];

  table.fields.forEach(field => {
    const { constraint } = field;

    if(constraint)
      statement.push(constraint);
  })

  if(statement.length)
    return `ALTER TABLE ${table.name} ${statement.join(", ")}`
}

function column(from: Field){
  if(from.datatype === undefined)
    return;

  let statement = `${Util.qualify(from.column)} ${from.datatype}`;

  if(!from.nullable)
    statement += " NOT NULL";

  if(from.default !== undefined)
    statement += ` DEFAULT ${Util.escapeString(from.default)}`;

  if(from.datatype == "INT" && from.increment)
    statement += " AUTO_INCREMENT";

  if(from.primary)
    statement += " PRIMARY KEY";

  return statement;
}