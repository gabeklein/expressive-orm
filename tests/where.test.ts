import { Entity, Int, One, Query, VarChar } from '../src';

class Foo extends Entity {
  name = VarChar();
  color = VarChar();
}

class Bar extends Entity {
  value = Int();
  color = VarChar();
}

it("will emit where clauses", () => {
  class Test extends Entity {
    a = Int();
    b = Int();
    c = Int();
    d = Int();
  }

  const query = new Query(where => {
    const test = where(Test);

    where(test.a).is(1);
    where(test.b).not(2);
    where(test.c).greater(3);
    where(test.d).less(4);
  });

  expect(query).toMatchSnapshot();
})

it("will assert a joined property's value", () => {
  const query = new Query(where => {
    const foo = where(Foo);
    const bar = where(Bar, { color: foo.color });

    where(bar.value).is(42);
  });
  
  expect(query).toMatchSnapshot();
})

it.skip("will assert an implictly joined value", () => {
  class Bar extends Entity {
    foo = One(Foo);
    greeting = "Hello World";
  }

  const query = new Query(where => {
    const bar = where(Bar);

    where(bar.foo.color).is("blue");
  });
  
  expect(query).toMatchInlineSnapshot();
})

it("will match values via object", () => {
  const query = new Query(where => {
    const foo = where(Foo);

    where(foo).has({
      name: "Gabe",
      color: "blue"
    })
  });

  expect(query).toMatchSnapshot();
})