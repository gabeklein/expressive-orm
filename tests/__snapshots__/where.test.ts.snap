// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will assert a joined property's value 1`] = `
	SELECT
	  COUNT(*)
	FROM
	  \`Foo\`
	  JOIN \`Bar\` ON \`Bar\`.\`color\` = \`Foo\`.\`color\`
	WHERE
	  \`Bar\`.\`value\` = 42
`;

exports[`will emit where clauses 1`] = `
	SELECT
	  COUNT(*)
	FROM
	  \`Test\`
	WHERE
	  \`Test\`.\`a\` = 1
	  AND \`Test\`.\`b\` <> 2
	  AND \`Test\`.\`c\` > 3
	  AND \`Test\`.\`d\` < 4
`;

exports[`will match values via object 1`] = `
	SELECT
	  COUNT(*)
	FROM
	  \`Foo\`
	WHERE
	  \`Foo\`.\`name\` = 'Gabe'
	  AND \`Foo\`.\`color\` = 'blue'
`;
